Add a File field, like for .pdf uploads

kienstra opened the pull request:
<!--- Please summarize this PR in the title above -->

# Changes
* Add a File field that allows non-image files:

<img width="823" alt="ioabsdifadsf" src="https://user-images.githubusercontent.com/4063887/125011796-bbdef280-e02e-11eb-9c59-378341d936c0.png">

* There's already an Image field, but that doesn't allow some files like `.pdf` files

Fixes [GF-3071](https://wpengine.atlassian.net/browse/GF-3071)

# Testing instructions
1. `cd /path/to/wp-content/plugins/genesis-custom-blocks/`

2. `git fetch`

3. `git checkout add/file-field`

4. `composer install && npm i && npm run build`

5. `/wp-admin` > Custom Blocks > Add New

6. Add a field with the type 'File': 
<img width="934" alt="file-here" src="https://user-images.githubusercontent.com/4063887/125011327-dd8baa00-e02d-11eb-8b1e-66031aae6bc7.png">

7. Click Publish

8. `/wp-admin` > Posts > Add New

9. Enter the name of the block you just created: 
<img width="850" alt="thebisdf" src="https://user-images.githubusercontent.com/4063887/125011464-23487280-e02e-11eb-8da4-b9c346db666e.png">

10. In the block, click 'Upload' and upload any `.pdf` file

11. Expected: The file name displays: 
<img width="823" alt="ioabsdifadsf" src="https://user-images.githubusercontent.com/4063887/125011659-7de1ce80-e02e-11eb-9f20-212210fa2508.png">

12. This won't preview the actual file in the Edit component, as the file could be a `.pdf`, a `.zip`, or any number of files

kienstra committed:
Rename useImage() to useMedia() to reuse it for the File field

kienstra committed:
Add a File field via PHP, and a test class

This still needs a JS component,
but it's registered with PHP now.

kienstra committed:
Add a React component for the new File field

Copied from the Image field,
including the test.

kienstra committed:
Fix failed PHPUnit test in TestBlockPost.php

kienstra committed:
Display the file name, not an <img> like the Image field

kienstra commented on js/src/block-editor/controls/file.js:
```diff
@@ -0,0 +1,121 @@
+/**
```
This file is mainly copied from `image.js`, so they could probably use a common component that they both call.

kienstra replied:
Actually, there are enough differences that a common component might make it more complex.

kienstra committed:
Add a validation method to echo/make a vailable the file

This means that the field
will actually display on the front-end.

kienstra committed:
Allow any file type for the file field

Like how the File block
accepts any file type:
https://github.com/WordPress/gutenberg/blob/2f32ec41c3663fa63eae3ed2494a25bda76ab36f/packages/block-library/src/file/edit.js#L205

kienstra committed:
Change the dashicon and text to File

This was mainly copied from the
Image control, so change this now.

kienstra committed:
Add the file field to the e2e test

Before, there was no e2e
test of this.

kienstra committed:
Add the testing .pdf file that I forgot to add

kienstra committed:
Fix an issue where the file doens't display on initial page load

Related to:
https://github.com/WordPress/gutenberg/issues/14623#issuecomment-499610251

kienstra committed:
Change some parameter names from image to ownMedia

They're not necessarily
images, as they could be files.

kienstra commented on package-lock.json:
```diff
@@ -5134,6 +5134,12 @@
+          "resolved": "https://registry.npmjs.org/wp-prettier/-/wp-prettier-2.2.1-beta-1.tgz",
+          "integrity": "sha512-+JHkqs9LC/JPp51yy1hzs3lQ7qeuWCwOcSzpQNeeY/G7oSpnF61vxt7hRh87zNRTr6ob2ndy0W8rVzhgrcA+Gw==",
+          "dev": true
+        },
```
Not related to this PR, but it's annoying to see this in the diff every time you do `npm install`

kienstra committed:
Simplify File::validate() to be one line

This should be a bit easier,
it was strange having $image_id.

kienstra committed:
Merge in develop, resolve conflict

There were only conflicts in tests/e2e/specs/all-fields.js

They were trival, I kept both.

kienstra committed:
Improve the styling of the image file name

Align it horizontally with the
help text.

kienstra commented on php/Blocks/Loader.php:
```diff
@@ -72,7 +72,7 @@ public function init() {
 	public function register_hooks() {
 		add_action( 'enqueue_block_editor_assets', [ $this, 'editor_assets' ] );
-		add_filter( 'block_categories', [ $this, 'register_categories' ] );
+		add_filter( 'block_categories_all', [ $this, 'register_categories' ] );
```
This will need to go out before WP `5.8` to prevent the PHP notice.

kienstra commented on js/src/block-editor/hooks/useMedia.js:
```diff
@@ -43,67 +43,64 @@ import { __, sprintf } from '@wordpress/i18n';
-		if ( newImage?.alt ) {
-			setImageAlt( newImage.alt );
-		} else if ( newImage?.source_url ) { // eslint-disable-line camelcase
+		if ( media?.alt ) {
```
Most of this file's diff is just renaming parameters. 

Like here, it changes `newImage` to `media`, as this hook is now used for files and images.

johnstonphilip commented:
Nice work Ryan! This is working well for me when logged in as an administrator. However, if I log in as a contributor, I see the file field, but it doesn't actually upload the file. This isn't a `must-fix` right now, but possibly something to add in the future, if the user doesn't actually have permission to upload files. Perhaps we shouldn't show the upload field to them. 

<img width="1378" alt="Screen Shot 2021-07-19 at 3 07 16 PM" src="https://user-images.githubusercontent.com/7538525/126213897-88e13a75-2ce5-4190-b018-c264a3c9387b.png">


The main thing I wanted to be sure of was user capabilities work, which they do. 

I was reading that you need to wrap `MediaUpload` in `MediaUploadCheck` here:
https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/media-upload/README.md

kienstra commented:
Thanks, @dreamwhisper! 

Great idea to test it with a contributor, never though of that. Also, good idea to use a `<MediaUploadCheck>`

kienstra committed:
Thanks to Jen's suggestion, use <MediaUploadCheck>

As Jen mentioned,
this can prevent the <MediaUpload> from showing
for users that can't upload files.

kienstra committed:
Merge in develop, resolve conflict

There was a trivial conflict in Loader.php.

kienstra committed:
Bump WP 'Tested up to' to 5.8, forgot earlier

This isn't the right PR to do it in,
but I forgot in #85.

kienstra commented:
If it's OK, I [bumped the WP Tested Up To version to 5.8](https://github.com/studiopress/genesis-custom-blocks/pull/74/commits/e798b9e4be98a42f6e9cbc46d726baf3ee6bde8c), forgot to do that in #85
